lms/static/scripts/frontend_apps/components/test/BasicLtiLaunchApp-test.js

```diff
@@ -254,11 +310,109 @@ describe('BasicLtiLaunchApp', () => {
+        viaCallbackUrl: 'https://lms.hypothes.is/api/files/1234',
+        sync: {
+          data: {
+            course: {
+              context_id: '12345',
+              custom_canvas_course_id: '101',
+            },
+          },
+          path: '/api/sync',
+        },
+      };
+      contentUrlCall = fakeApiCall.withArgs({
+        authToken: 'dummyAuthToken',
+        path: 'https://lms.hypothes.is/api/files/1234',
+      });
+
+      groupsCall = fakeApiCall.withArgs({
+        authToken: 'dummyAuthToken',
+        path: '/api/sync',
+        data: {
+          course: {
+            context_id: '12345',
+            custom_canvas_course_id: '101',
+          },
+        },
+      });
+    });
+
+    it('renders the spinner until both groups and contentUrl requests finish', async () => {
+      const contentUrl = contentUrlCall.resolves({
+        via_url: 'https://via.hypothes.is/123',
+      });
+      const groups = groupsCall.resolves(['group1', 'group2']);
+      const wrapper = renderLtiLaunchApp();
+      await contentUrl;
+      // Spinner should not go away after first request
+      wrapper.update();
+      assert.isTrue(wrapper.find('Spinner').exists());
+      await groups;
+      // Spinner should go away after the second request
+      await waitForElementToRemove(wrapper, 'Spinner');
+      // iframe should be visible
+      assert.equal(wrapper.find('iframe').prop('style').visibility, 'visible');
+    });
+
+    it('renders the iframe hidden after contentUrl succeeds and groups remains pending', async () => {
+      const contentUrl = contentUrlCall.resolves({
+        via_url: 'https://via.hypothes.is/123',
+      });
+      const wrapper = renderLtiLaunchApp();
+      await contentUrl;
+      wrapper.update();
+      assert.equal(wrapper.find('iframe').prop('style').visibility, 'hidden');
+    });
+
+    it('shows an error dialog if the first request fails and second succeeds', async () => {
+      const contentUrl = contentUrlCall.rejects(new ApiError(400, {}));
+      const groups = groupsCall.resolves(['group1', 'group2']);
+      const wrapper = renderLtiLaunchApp();
+      await contentUrl;
+      // Should show an error after the first request fails
+      await waitForElement(wrapper, 'FakeDialog[title="Authorize Hypothesis"]');
+      await groups;
+      // Should still show an error even if the second request does not fail
+      await waitForElement(wrapper, 'FakeDialog[title="Authorize Hypothesis"]');
+    });
+
+    it('shows an error dialog if the first request succeeds and second fails', async () => {
+      const contentUrl = contentUrlCall.resolves({
+        via_url: 'https://via.hypothes.is/123',
+      });
+      const groups = groupsCall.rejects(new ApiError(400, {}));
+      const wrapper = renderLtiLaunchApp();
+      await contentUrl;
+      // Should not show an error yet
+      assert.isFalse(
+        wrapper.find('FakeDialog[title="Authorize Hypothesis"]').exists()
+      );
```

Comment:
Asserting that something is not rendered "yet" is a bit tricky because it is possible that some batched state updates haven't yet been flushed (Preact currently does this internally using something equivalent to `queueMicrotask`). The pattern I would probably use is something like:

1. Put the `groups` Promise into a pending state where it won't resolve or reject
2. Wait one macro-task (`setTimeout` with a delay of 0). This will let any pending renders flush.
3. Check that the auth dialog hasn't rendered
4. Reject the `groups` promise
5. Now wait for the "Authorize Hypothesis" dialog to appear

Reply:
Also, be aware that you need to call `wrapper.update(...)` again before querying it here.